File: Graph/BFS_distance.cpp
/*
BFS in Graph 
연결 그래프에서 1번 노드와의 거리 
*/
#include <bits/stdc++.h>
using namespace std;

int v, e;   // node, edge
vector<int> adj[10];
bool dist[10];  // 거리 저장

void BFS(){
    fill(dist, dist+10, -1);    // 거리 초기화 : 방문하지 않은 곳 '-1'
    queue<int> Q;
    // 1. 시작 노드를를 넣고, 거리 = 0
    Q.push(1);      
    dist[1] = 0;
    // 2. Queue가 빌때까지, 연결된 노드(인접리스트)을 Queue에 넣고, 방문(거리)표시 
    while(!Q.empty()){
        int cur = Q.front();
        Q.pop();
        for (auto nxt : adj[cur]){
            if (dist[nxt] != -1) continue;
            Q.push(nxt);
            dist[nxt] = dist[cur] + 1;   // 거리 += 1 증가 
        }
    }
}

File: Graph/BFS_distance.py
num_node = 6 
edge = [(1,2), (5,1), (3,1), (4,6), (4,3), (2,5)]

adj = [[] for _ in range(num_node+1)]   # 1부터 시작 
for i in range(len(edge)):
    #u, v = map(int, input().split())
    u, v = edge[i]
    # No direction
    adj[u].append(v)    # u->v
    adj[v].append(u)    # v->u

# sorting : 만약 크기순으로 BFS가 필요하다면 adj list sorting 필요 
for l in adj:
    l.sort()
        
print("Adjacent list : {}".format(adj))

# visited list = fill(dist, dist+num_node, -1)
dist = [-1]*(num_node+1) # 거리 초기화 : 방문하지 않은 곳 -1

from collections import deque
def BFS(st):
    Q = deque()
    Q.append(st)
    dist[st] = 0 
    
    while (Q):
        cur = Q.popleft()   # Q.front(); Q.pop();
        for nxt in adj[cur]:
            if (dist[nxt] != -1):
                continue
            Q.append(nxt)
            dist[nxt] = dist[cur] + 1
            
    return dist

print(BFS(1))
